File: bufferring/common/ptable.py
class ProgressTable:
    def __init__(self, size, rank, name_list):
        assert size > 0
        assert rank >= 0 and rank < size
        self.__table = size * [{name: 0 for name in name_list}]
        self.__size = size
        self.__rank = rank

    def get_my_cycle(self, name):
        return self.__table[self.__rank][name]

    def update(self, node, name):
        assert node >= 0 and node < self.__size
        self.__table[node][name] += 1

    def blocked(self, thresh):
        maximum = max([max(d.values()) for d in self.__table])
        minimum = min([min(d.values()) for d in self.__table])

        if thresh > maximum - minimum:
            return False
        elif minimum in self.__table[self.__rank].values():
            return False
        else:
            return True

File: bufferring/core/ptable.cpp
#include "ptable.hpp"

ProgressTable::ProgressTable(int size, int rank, int names_len) : size_(size), rank_(rank), names_len_(names_len)
{
	table = new int *[names_len];
	for (int i = 0; i < names_len; i++)
		table[i] = new int[size]();
}

ProgressTable::~ProgressTable(void)
{
	for (int i = 0; i < names_len_; i++)
		delete []table[i];
	delete []table;
}

void ProgressTable::update(int node, int nametag)
{
	table[nametag][node]++;
}

bool ProgressTable::blocked(int nametag, int thresh)
{
	int max, min;
	max = min = table[nametag][0];

	for (int i = 1; i < size_; i++) {
		if (max < table[nametag][i])
			max = table[nametag][i];
		if (min > table[nametag][i])
			min = table[nametag][i];
	}

	if (thresh > max - min)
		return false;
	else if (table[nametag][rank_] == min)
		return false;
	else
		return true;
}
